### PR TITLE
Add user friendly description to internal delta queries

### PR DESCRIPTION
**Pull request by ekoifman**
to make Spark UI easier to follow
Delta operation generates multiple queries to manage the transaction log.
For example
```
val data = spark.range(0, 15)
data.write.format("delta").save("/tmp/delta-table")
val df = spark.read.format("delta").load("/tmp/delta-table")

import io.delta.tables._
import org.apache.spark.sql.functions._
val deltaTable = DeltaTable.forPath("/tmp/delta-table")
deltaTable.delete(condition = expr("id == 0"))
```

executes several "housekeeping" queries.  This enhancement makes it easier to figure out which is which.

"Default Names" and "Enhanced Names" screen shots show the before and after UI.  (Stack traces shown in the UI are not useful here)

"Default Names"
DefaultNames

"Enhanced Names"
Enhanced Names

Closes https://github.com/delta-io/delta/issues/528

### CODE REVIEW

**Comment by ekoifman** [comment]
Hi,
@tdas @brkyvz any chance you can look at this?

**Comment by brkyvz** [comment]
Thanks @ekoifman I plan to take a look by tomorrow the latest. Sorry for the delay

File: src/main/scala/org/apache/spark/sql/delta/Snapshot.scala [hunk 1]
```diff
@@ -136,46 +136,48 @@ class Snapshot(
...
-    val stateReconstructionCheck = spark.sessionState.conf.getConf(
-      DeltaSQLConf.DELTA_STATE_RECONSTRUCTION_VALIDATION_ENABLED)
-    if (_computedState.protocol == null) {
-      recordDeltaEvent(
-        deltaLog,
-        opType = "delta.assertions.missingAction",
-        data = Map("version" -> version.toString, "action" -> "Protocol", "source" -> "Snapshot"))
-      if (stateReconstructionCheck) {
-        throw DeltaErrors.actionNotFoundException("protocol", version)
+  protected lazy val computedState: State = withJobDescription("Snapshot read", spark) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/Snapshot.scala:139]: withJobDescription(`Compute snapshot for version: $version`) {

File: src/main/scala/org/apache/spark/sql/delta/Snapshot.scala [hunk 2]
```diff
@@ -136,46 +136,48 @@ class Snapshot(
...
+    {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/Snapshot.scala:140]: you don't need a second `{}`. The closure above should be sufficient

File: src/main/scala/org/apache/spark/sql/delta/PartitionFiltering.scala [hunk 3]
```diff
@@ -31,10 +31,12 @@ trait PartitionFiltering { self: Snapshot =>
       DeltaTableUtils.splitMetadataAndDataPredicates(filter, metadata.partitionColumns, spark)._1
     }
 
-    val files = DeltaLog.filterFileList(
-      metadata.partitionSchema,
-      allFiles.toDF(),
-      partitionFilters).as[AddFile].collect()
+    val files = withJobDescription("partition filter-files", spark) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/PartitionFiltering.scala:34]: nit: don't think you need the hyphen between `filter-files`

File: src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala [hunk 4]
```diff
@@ -136,11 +136,13 @@ case class DeleteCommand(
               if (numTouchedFiles == 0) {
                 Array.empty[String]
               } else {
-                data
-                  .filter(new Column(cond))
-                  .filter(deletedRowUdf())
-                  .select(new Column(InputFileName())).distinct()
-                  .as[String].collect()
+                withJobDescription("Delete filter-files", sparkSession) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala:139]: `Finding files to delete`?

File: src/main/scala/org/apache/spark/sql/delta/PartitionFiltering.scala (hunk 3 above)
Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/PartitionFiltering.scala:34]: how about `Filtering files for query`?

File: src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala [hunk 5]
```diff
@@ -164,7 +166,9 @@ case class DeleteCommand(
 
             val rewrittenFiles = withStatusCode(
               "DELTA", s"Rewriting ${filesToRewrite.size} files for DELETE operation") {
-              txn.writeFiles(updatedDF)
+              withJobDescription("Delete write-data", sparkSession) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala:169]: `Writing files without deleted records`

File: src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala [hunk 6]
```diff
@@ -207,13 +207,19 @@ case class MergeIntoCommand(
 
       val deltaActions = {
        if (isSingleInsertOnly && spark.conf.get(DeltaSQLConf.MERGE_INSERT_ONLY_ENABLED)) {
-         writeInsertsOnlyWhenNoMatchedClauses(spark, deltaTxn)
+         withJobDescription("MergeInto write-data-insert", spark) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala:210]: `Writing inserted data with merge`

File: src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala [hunk 7]
```diff
@@ -207,13 +207,19 @@ case class MergeIntoCommand(
...
+           writeInsertsOnlyWhenNoMatchedClauses(spark, deltaTxn)
+         }
        } else {
          val filesToRewrite =
            recordDeltaOperation(targetDeltaLog, "delta.dml.merge.findTouchedFiles") {
-             findTouchedFiles(spark, deltaTxn)
+             withJobDescription("MergeInto filter-files", spark) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala:216]: Filtering files for merge

File: src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala [hunk 8]
```diff
@@ -207,13 +207,19 @@ case class MergeIntoCommand(
...
+               findTouchedFiles(spark, deltaTxn)
+             }
            }
-         val newWrittenFiles = writeAllChanges(spark, deltaTxn, filesToRewrite)
+         val newWrittenFiles = withJobDescription("MergeInto write-data", spark) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala:220]: `Writing merged data`

File: src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala [hunk 9]
```diff
@@ -129,10 +129,12 @@ case class UpdateCommand(
       }.asNondeterministic()
       val filesToRewrite =
         withStatusCode("DELTA", s"Finding files to rewrite for UPDATE operation") {
-          data.filter(new Column(updateCondition))
-            .filter(updatedRowUdf())
-            .select(input_file_name())
-            .distinct().as[String].collect()
+          withJobDescription("Update filter-files", sparkSession) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:132]: `Filtering files for update`

Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:132]: Can we add a description for the jobs in the other cases above?

File: src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala [hunk 10]
```diff
@@ -149,8 +151,10 @@ case class UpdateCommand(
           removeFilesFromPaths(deltaLog, nameToAddFile, filesToRewrite, operationTimestamp)
         val rewrittenFiles =
           withStatusCode("DELTA", s"Rewriting ${filesToRewrite.size} files for UPDATE operation") {
-            rewriteFiles(sparkSession, txn, tahoeFileIndex.path,
-              filesToRewrite, nameToAddFile, updateCondition)
+            withJobDescription("Update write-files", sparkSession) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:154]: `Writing files with updated records`?

File: src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala [hunk 11]
```diff
@@ -102,4 +103,23 @@ trait DeltaLogging
...
+  def withJobDescription[U](desc: String, session: SparkSession)(body: => U): U = {
+    val sc = session.sparkContext
+    val oldDesc = sc.getLocalProperty(SparkContext.SPARK_JOB_DESCRIPTION)
+    val suffix = if (oldDesc == null) {
+      ""
+    } else {
+      s" $oldDesc:"
+    }
+    try {
+      sc.setJobDescription(s"Delta:$suffix $desc")
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala:119]: In a nested invocation, this may seem a bit weird. Do we need to include the parent invocation's description?

**Review by brkyvz — CHANGES_REQUESTED** [review]
Thank you so much for this huge usability improvement @ekoifman. I left some suggestions. Let me know what you think

File: src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala (hunk 11 above)
Comment (ekoifman) [reply src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala:119]: Thanks for the feedback @brkyvz 
In an app where many Delta based queries are submitted, this helps correlate the "housekeeping" queries with the top level user query (in Spark 3.0 jobDescription is the SQL text).
In Delta code I made the change such that `withJobDescription` calls are not nested.  So I think the `suffix` is useful.

The other comments make sense, I'll make the adjustments.

Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala:119]: ah got it. It's not nested with respect to Delta but something else.

File: src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala [hunk 12]
```diff
@@ -129,10 +135,12 @@ case class UpdateCommand(
       }.asNondeterministic()
       val filesToRewrite =
         withStatusCode("DELTA", s"Finding files to rewrite for UPDATE operation") {
-          data.filter(new Column(updateCondition))
-            .filter(updatedRowUdf())
-            .select(input_file_name())
-            .distinct().as[String].collect()
+          withJobDescription("Filtering files for update", sparkSession) {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:138]: Actually now that I realize `withStatusCode`, shall we just merge `withStatusCode` and `withJobDescription`?

Comment (ekoifman) [reply src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:138]: They don't always match 1-1
there are at least some places where `withStatusCode` is not called, eg.
`PartitionFiltering`
`  protected lazy val computedState: State =` in `Snapshot`
And other places like
`GenerateSymlinkManifestImpl.recordManifestGeneration` where `withStatusCode` is "far" from Spark query.

But mainly, I think this may cause calls to nest, which is not desirable for `withJobDescription`

Comment (ekoifman) [reply src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:138]: @brkyvz do think this requires anything else?

Comment (brkyvz) [reply src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala:138]: Yeah, I just think that having one mechanism to state what's going on is actually pretty useful. Otherwise developers may be confused when to use `withJobDescription` versus `withJobDescription`. I feel it'd be a good idea to merge the two. If we do merge the two, it'd be great to just add the `withStatusCode` to places where you've already added the `withJobDescription` (if they don't already exist)

File: src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala [hunk 13]
```diff
@@ -18,10 +18,11 @@ package org.apache.spark.sql.delta.metering
 
 import scala.util.Try
 import scala.util.control.NonFatal
-
 import com.databricks.spark.util.{DatabricksLogging, OpType, TagDefinition}
 import com.databricks.spark.util.MetricDefinitions.{EVENT_LOGGING_FAILURE, EVENT_TAHOE}
 import com.databricks.spark.util.TagDefinitions.{TAG_OP_TYPE, TAG_TAHOE_ID, TAG_TAHOE_PATH}
+import org.apache.spark.SparkContext
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala:24]: unnecessary changes?

File: src/main/scala/org/apache/spark/sql/delta/util/DeltaProgressReporter.scala [hunk 14]
```diff
@@ -27,8 +29,30 @@ trait DeltaProgressReporter extends Logging {
       defaultMessage: String,
       data: Map[String, Any] = Map.empty)(body: => T): T = {
     logInfo(s"$statusCode: $defaultMessage")
-    val t = body
+    val t = withJobDescription(defaultMessage)(body)
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/util/DeltaProgressReporter.scala:32]: Thank you for doing this!

File: src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala (hunk 13 above)
Comment (ekoifman) [reply src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala:24]: yep.  I'll update

File: src/main/scala/org/apache/spark/sql/delta/util/DeltaProgressReporter.scala (hunk 14 above)
Comment (ekoifman) [reply src/main/scala/org/apache/spark/sql/delta/util/DeltaProgressReporter.scala:32]: @brkyvz do you have any more concerns?

File: src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala [hunk 15]
```diff
@@ -18,7 +18,6 @@ package org.apache.spark.sql.delta.metering
 
 import scala.util.Try
 import scala.util.control.NonFatal
-
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala:21]: can you remove this spurious change please

**Review by brkyvz — APPROVED** [review]
LGTM

File: src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala (hunk 15 above)
Comment (ekoifman) [reply src/main/scala/org/apache/spark/sql/delta/metering/DeltaLogging.scala]: done
